### ISSUE TITLE
There is a problem with the ROM size occupied by the esp8266 / esp32 calculation program.

### ISSUE THREAD

**Issue by xiezhoubin** [Sloeber/arduino-eclipse-plugin#1476]
Will exceed the actual size.

![image](https://user-images.githubusercontent.com/75734559/169683772-faf0520d-e3f2-4fda-9885-692b24e3d23a.png)
![QQ图片20220522153454](https://user-images.githubusercontent.com/75734559/169684089-462d83fc-2dd8-47f6-8054-cf79adffca11.png)

**Comment by xiezhoubin** [comment]
![image](https://user-images.githubusercontent.com/75734559/169684145-58589b34-a0d8-40b6-8c5d-d51eb47588fe.png)
![image](https://user-images.githubusercontent.com/75734559/169684187-3f757650-05b3-4d49-b9b5-40762eb47570.png)

**Comment by xiezhoubin** [comment]
Can this problem be fixed?

**Comment by jantje** [comment]
This arduino way size calculation is pretty new and only tested with avr boards.
As it involves some higher levels of magic -and so does esp stuff- I'm actually surprised the results are so close.
I'll see if I can reproduce and fix.

**Comment by xiezhoubin** [comment]
I hope you can fix this bug. Thank you very much.

**Comment by jantje** [comment]
I looked into this issue and it seems the esp8266 does something which I do not understand.
To get the size the arduino way Sloeber runs the command 
${recipe.size.pattern}
The output of this command is parsed with the regular expressions (in a awk file) 
${recipe.size.regex}
${recipe.size.regex.data}
${recipe.size.regex.eeprom}
When I ran a test with the board below and the default ino sketch
![afbeelding](https://user-images.githubusercontent.com/1175841/170147436-be6aca05-73ef-47ee-8095-446e534809f3.png)
${recipe.size.pattern} resulted into 
`"C:\eclipse\arduinoPlugin\packages\esp8266\tools\xtensa-lx106-elf-gcc\3.0.4-gcc10.3-1757bed/bin/xtensa-lx106-elf-size" -A "L:\test\runtime-EclipseApplication_nl\test\Release/test.elf" 
`
When running that command I get following result.
![afbeelding](https://user-images.githubusercontent.com/1175841/170147703-7a82c91e-7646-493a-aa17-a307225556ac.png)

The first field in the result is calculated with the regex ${recipe.size.regex} this regex is defined in platform.txt as
recipe.size.regex=^(?:\.irom0\.text|\.text|\.text1|\.data|\.rodata|)\s+([0-9]+).*
Sloeber interprets this as: for "all the lines that hold .irom0.text or .text or .text1 or .data or .rodata and one or more numbers" sum the second column
If you look at the output of the command you'll see that .irom0.text is 231500 which is already way bigger than the value Arduino IDE provides.
I tried to look how I could get the 10644 reported by Arduino IDE but I did not find a way.

So basically I do not know how to calculate the values. Which fields are needed and how do I do the calculation with these fields?

Please provide input on how to calculate the required values


PS sloeber regular expression behaviour is different from Arduino IDE behaviour as Sloeber uses awk and I assume Arduino IDE uses java regex
PPS I also notice arduino IDE does not provide a second line. I assume that is because ${recipe.size.regex.eeprom} is not defined.

**Comment by jantje** [comment]
oeps problem in previus post. Arduino IDE says 260089 not 10644 
If I add  .irom0.text .text .text1 .data  .rodata I get 260089 
So something is wrong when converting from java regex to awk

**Comment by jantje** [comment]
I have found the root cause 😄 

### Workaround
1) Replace the following line in the platform.txt
`recipe.size.regex=^(?:\.irom0\.text|\.text|\.text1|\.data|\.rodata|)\s+([0-9]+).*`
with
`recipe.size.regex=^(?:\.irom0\.text|\.text|\.text1|\.data|\.rodata)\s+([0-9]+).*`

2) delete the platform.sloeber.txt file

3) delete the build folders of all the project configs (default is Release) that need the workaround

### Explanation
The | means or and the | at the end of the list is apparently ignored by java regex but the sloeber converted AWK command results in "all lines are good lines".
As the output already contains the total the result is 2 times total

**Comment by xiezhoubin** [comment]
> 我找到了根本原因😄
> 
> ### 解决方法
> 1. 将 platform.txt 中的以下行替换
>    `recipe.size.regex=^(?:\.irom0\.text|\.text|\.text1|\.data|\.rodata|)\s+([0-9]+).*`
>    为
>    `recipe.size.regex=^(?:\.irom0\.text|\.text|\.text1|\.data|\.rodata)\s+([0-9]+).*`
> 2. 删除 platform.sloeber.txt 文件
> 3. 删除需要解决方法的所有项目配置的构建文件夹（默认为 Release）
> 
> ### 解释
> 该| 表示或和| java regex 显然忽略了列表末尾的内容，但 sloeber 转换的 AWK 命令导致“所有行都是好行”。 由于输出已经包含总数，因此结果是总数的 2 倍

This method is feasible. Will awk be updated in the next version so that there is no need to modify the package file, otherwise it is still a little troublesome.
![image](https://user-images.githubusercontent.com/75734559/170158718-0ce2485e-0249-4c44-9a4e-4cfa400bee2e.png)
![image](https://user-images.githubusercontent.com/75734559/170158759-0b575358-8846-4032-914b-ef164d5e904a.png)

**Comment by jantje** [comment]
I already made a fix but it needs to be tested before I release it and as there is not automated test for this .....
So it is comming

**Comment by xiezhoubin** [comment]
> I already made a fix but it needs to be tested before I release it and as there is not automated test for this ..... So it is comming

OK, esp32 also has such a calculation error and needs to be repaired.

**Comment by jantje** [comment]
Though the build fails this should be fixed.
I was not aware the test I changed was part of the regression testing and now it is build to fail when things change so I can see only the things changed I wanted to change :-(

**Comment by jantje** [comment]
Should be fixed in the nightly
